src/EPR.SubsidiaryBulkUpload.Application/Services/CustomCsvReader.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System.Globalization;
+using CsvHelper;
+using CsvHelper.Configuration;
+
+namespace EPR.SubsidiaryBulkUpload.Application.Services;
+
+public class CustomCsvReader : CsvReader
+{
+    public CustomCsvReader(IParser parser)
+        : base(parser)
+    {
+    }
+
+    public CustomCsvReader(TextReader reader, IReaderConfiguration configuration)
+        : base(reader, configuration)
+    {
+    }
+
+    public CustomCsvReader(TextReader reader, CultureInfo culture, bool leaveOpen = false)
+        : base(reader, culture, leaveOpen)
+    {
+    }
+
+    public virtual MissingFieldFound MissingFieldMappingFound { get; set; }
+
+    protected override void ValidateHeader(ClassMap map, List<InvalidHeader> invalidHeaders)
+    {
+        base.ValidateHeader(map, invalidHeaders);
+
+        // We'll only run our validation if the base validation did not find any problems
+        if (invalidHeaders.Count == 0)
+        {
+            var unexpectedHeaders = new List<string>();
+            for (var i = 0; i < HeaderRecord.Length; i++)
+            {
+                var header = HeaderRecord[i];
+                if (!isHeaderMapped(map, header, i))
+                {
+                    unexpectedHeaders.Add(header);
+                }
+            }
+
+            if (unexpectedHeaders.Count != 0)
+            {
+                throw new UnexpectedHeadersException(Context, unexpectedHeaders);
```

Comment:
This should not be thrown.
We agreed to set a property value and use that instead of throwing unhandled exceptions. This makes the class really difficult to use.

Reply:
exception removed